### PR TITLE
feat: integrate image generation with cost tracking and metrics (#501)

### LINKED ISSUE #501: Image Generation: Integrate with cost tracking and metrics
## Problem

llm4s tracks costs and metrics for LLM completion calls via `MetricsCollector` and `TraceEvent.CostRecorded`, but image generation operations are not tracked. Image generation has significant per-request costs (e.g., OpenAI GPT Image: $0.02-$0.19 per image depending on quality/size).

**Complexity: 2/5 (Low-Medium)** — Integration work following existing patterns in `MetricsRecording` and `TraceEvent`. Good first issue.

### 1. Emit cost trace events for image generation

After each image generation call, emit a `TraceEvent.CostRecorded` with:
- Model name
- Operation: `"image_generation"` or `"image_edit"`
- Cost estimate based on model and parameters (size, quality, count)

### 2. Record metrics via MetricsCollector

Increment counters and record histograms for:
- `image_generation_total` — count of generation requests
- `image_generation_duration_seconds` — latency histogram
- `image_generation_errors_total` — error count by type
- `image_generation_cost_usd` — estimated cost

### 3. Add pricing data

Add image generation pricing to `ModelRegistry` or a new `ImagePricingRegistry`:

| Model | Size | Quality | Price per image |
|-------|------|---------|-----------------|
| gpt-image-1 | 1024x1024 | medium | $0.04 |
| gpt-image-1 | 1024x1024 | high | $0.08 |
| gpt-image-1-mini | 1024x1024 | medium | $0.02 |
| dall-e-3 | 1024x1024 | standard | $0.04 |
| dall-e-3 | 1024x1024 | hd | $0.08 |

(Pricing from OpenAI as of 2026. Other providers would need their own pricing tables.)

## Files likely affected

- `ImageGenerationClient` implementations (emit events after calls)
- `TraceEvent` (may need new event type or reuse `CostRecorded`)
- `MetricsCollector` (add image-specific recording methods)
- New: `ImagePricingRegistry` or extend `ModelRegistry`

## Definition of Done

- [ ] Cost estimates emitted as trace events after image generation calls
- [ ] Metrics recorded (count, latency, errors, cost)
- [ ] Pricing data for OpenAI image models
- [ ] Tests verifying metrics emission
- [ ] Console tracing displays image generation costs

### PR DESCRIPTION
**Pull request by kallal79**
- Add ImagePricingRegistry with per-image pricing for DALL-E 2/3, GPT-Image-1, Stable Diffusion XL, and Stability AI models across quality/size variants
- Add InstrumentedImageGenerationClient decorator that wraps any ImageGenerationClient with metrics collection and trace event emission
- Add observeImageGeneration and recordImageGenerationCost to MetricsCollector
- Add Prometheus counters (llm4s_image_generations_total, llm4s_images_generated_total, llm4s_image_generation_cost_usd_total) and histogram (llm4s_image_generation_duration_seconds) to PrometheusMetrics
- Add ImageGenerationCompleted TraceEvent with JSON serialization
- Update ConsoleTracing, LangfuseTracing, OpenTelemetryTracing, and TraceCollector to handle the new trace event
- Add ImageGeneration.client(config, metrics, tracing) factory overload while preserving backward compatibility of client(config)
- Add comprehensive tests: ImagePricingRegistrySpec (13 tests) and ImageGenerationCostTrackingSpec (15 tests)

Closes #501

### CODE REVIEW

File: modules/core/src/main/scala/org/llm4s/imagegeneration/InstrumentedImageGenerationClient.scala [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    val outcome = result match {
+      case Right(_) => Outcome.Success
+      case Left(_)  => Outcome.Error(ErrorKind.Unknown)
+    }
+
+    metrics.observeImageGeneration(providerName, model, operation, outcome, durationFD, imageCount)
+
+    val costUsd = result match {
+      case Right(_) =>
+        val cost = ImagePricingRegistry.estimateCost(model, options.quality, options.size, imageCount)
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/imagegeneration/InstrumentedImageGenerationClient.scala]: Cost is computed from options.size, but provider billing can use a different normalized size.
Example: OpenAI maps Landscape768x512 to 1792x1024 for dall-e-3.
Please estimate cost from the actual API size/model payload, not only requested size metadata.
Guideline principle: billing metrics must use the real charged dimensions.

File: modules/core/src/main/scala/org/llm4s/imagegeneration/InstrumentedImageGenerationClient.scala [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+      size = sizeStr,
+      quality = quality,
+      durationMs = durationMs,
+      costUsd = costUsd,
+      success = result.isRight,
+      errorMessage = result.left.toOption.map(_.message)
+    )
+    tracing.traceEvent(event)
+
+    costUsd.foreach(c => tracing.traceCost(c, model, "image_generation", 0, "image_generation"))
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/imagegeneration/InstrumentedImageGenerationClient.scala]: This always records operation as image_generation even for edit flow.
Please pass operation specific value so cost traces are accurate for generate vs edit.
Guideline principle: telemetry labels must reflect real runtime operation type.

File: modules/core/src/main/scala/org/llm4s/metrics/PrometheusMetrics.scala [hunk 3]
```diff
@@ -144,6 +177,62 @@ final class PrometheusMetrics(
...
+      imageGenerationDuration.labelValues(provider, model, operation).observe(duration.toMillis / 1000.0)
+
+      if (outcome == Outcome.Success && imageCount > 0) {
+        imagesGeneratedTotal.labelValues(provider, model).inc(imageCount.toDouble)
+      }
+
+      outcome match {
+        case Outcome.Error(errorKind) =>
+          val errorLabel = errorKind.toString.toLowerCase
+          errorsTotal.labelValues(provider, errorLabel).inc()
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/metrics/PrometheusMetrics.scala:220]: Issue #501 asks for image_generation_errors_total, but errors are only added to global llm4s_errors_total here.
Please add dedicated image error counter or update issue scope and PR text explicitly.
Guideline principle: implementation should match declared acceptance metrics.

**Review by vim89 — CHANGES_REQUESTED** [review]
Request changes.
This PR brings strong value for observability and cost tracking, and the architecture direction fits core image generation work.
But cost estimation currently uses requested size metadata, not provider normalized billed size, so cost can be wrong for models like dall-e-3.
Cost tracing also labels edit operations as image_generation, and the issue asked for image_generation_errors_total but code records only the generic error counter.
Please fix these gaps before merge.
I did not run tests/build.

**Review by rorygraves — CHANGES_REQUESTED** [review]
## PR Review: #818 — Integrate image generation with cost tracking & metrics

**Status: Changes Requested**

Good architectural approach — the decorator pattern (`InstrumentedImageGenerationClient`) is clean, the `ImagePricingRegistry` is well-structured, and the integration across all 4 tracing backends (Console, Langfuse, OTel, TraceCollector) is thorough. Test coverage is strong (539 lines). The backward-compatible factory overload is well done.

### Issues Found

| # | File | Line | Severity | Issue | Suggestion |
|---|------|------|----------|-------|------------|
| 1 | `InstrumentedImageGenerationClient.scala` | 345 | **High** | **All errors map to `ErrorKind.Unknown`.** `ImageGenerationError` has 7 specific subtypes (`AuthenticationError`, `RateLimitError`, `ServiceError`, `ValidationError`, etc.) but the outcome is always `Outcome.Error(ErrorKind.Unknown)`. This makes error metrics useless for diagnosis — you can't distinguish auth failures from rate limits in Prometheus. | Add a mapping: `AuthenticationError → ErrorKind.Authentication`, `RateLimitError → ErrorKind.RateLimit`, `ServiceError → ErrorKind.ServiceError`, `ValidationError → ErrorKind.Validation`, etc. |
| 2 | `InstrumentedImageGenerationClient.scala` | 339, 352 | **High** | **Cost estimated from `options.size` (requested), not actual billed size.** For dall-e-3, OpenAI normalizes some sizes (e.g., `Landscape768x512` doesn't exist — it maps to `1792x1024` for billing). The lookup uses the *requested* `ImageSize` enum, not the provider's actual billed dimensions. For unmapped sizes it returns `None` (silent no-cost). | Document this limitation. Consider ensuring the pricing table covers all `ImageSize` enum values per model, or use generated image metadata if available. |
| 3 | `InstrumentedImageGenerationClient.scala` | 375 | **Medium** | **`traceCost` called with `tokenCount = 0`.** `traceCost(c, model, "image_generation", 0, "image_generation")` hardcodes `tokenCount = 0`. The `CostRecorded` event was designed for LLM token costs — recording 0 tokens pollutes token-based cost analytics. | Either skip `traceCost` entirely (the `ImageGenerationCompleted` event already carries cost), or add a dedicated image cost trace method. |
| 4 | `PrometheusMetrics.scala` | 488-506 | **Medium** | **Error label format inconsistency.** The `status` label uses `s"error_$snakeCase"` (regex conversion), while `errorsTotal` uses `errorKind.toString.toLowerCase`. Two different label formats for the same error across counters. | Standardize via a shared helper for `ErrorKind` → label conversion. |
| 5 | `MetricsCollector.scala` | — | **Medium** | **Missing `image_generation_errors_total` counter.** Issue #501 specifies this counter. Errors are recorded only via the generic `errorsTotal`, making it impossible to filter image-specific errors in dashboards. | Add a dedicated counter or explicitly document that it's deferred. |
| 6 | `InstrumentedImageGenerationClient.scala` | 278-279 | **Medium** | **Edit operation hardcodes `ImageSize.Square512` as default.** When `ImageEditOptions.size` is `None`, cost lookup uses `512x512`. But the actual billed size depends on the provider's default (e.g., `1024x1024` for dall-e-3 edits). | Use a provider-aware default, or skip cost estimation when size is unknown rather than guessing. |
| 7 | `InstrumentedImageGenerationClient.scala` | 334 | **Low** | **`@unused requestedCount` parameter.** On failure `imageCount = 0`. The trace event only shows actual count, not requested count — losing debug info for partial failures. | Either remove the parameter or include it in the trace event. |
| 8 | `ImagePricingRegistry.scala` | 109-136 | **Low** | **Static pricing will go stale.** Acceptable for a first pass but add a comment noting when prices were last verified. | The ScalaDoc partially covers this — consider adding a "last verified" date. |

### Positive Notes
- Clean decorator pattern — wraps any client without modifying implementations
- Backward-compatible factory: `client(config)` still works, `client(config, metrics, tracing)` adds instrumentation
- All 4 tracing backends updated consistently
- `ImagePricingRegistry` well-tested with edge cases
- `PrometheusMetrics` uses `Try/recover` consistent with existing code
- Tests cover success/failure paths, cost recording, trace events, and noop compatibility

### Summary
2 high, 4 medium, 2 low severity issues. The **high** issues are: all errors mapping to `ErrorKind.Unknown` (makes error metrics useless) and cost estimation using requested size rather than actual billed size.

**Comment by kallal79** [comment]
HI @rorygraves , @vim89 , Requesting re‑review after addressing cost tracking, operation labels, error subtype mapping, and adding image_generation_errors_total per issue #501.

**Comment by kallal79** [comment]
HI @rorygraves , @vim89 , Requesting re‑review after addressing cost tracking, operation labels, error subtype mapping, and adding image_generation_errors_total per issue https://github.com/llm4s/llm4s/issues/501.

**Comment by kallal79** [comment]
HI @rorygraves , @vim89 , Requesting re‑review after addressing cost tracking, operation labels, error subtype mapping, and adding image_generation_errors_total per issue https://github.com/llm4s/llm4s/issues/501.

**Comment by kallal79** [comment]
Thanks @rorygraves and @vim89 for the thorough review and merge — glad the image generation cost tracking and metrics integration is now complete in main. Appreciate the collaboration!
